### PR TITLE
Add support for __slots__-optimized classes

### LINKED ISSUE #57: Add support of __slots__ classes
Currently we support a default encoding of classes using their underlying `__dict__`. This does not cover all classes as some classes are optimized to use `__slots__` instead of having a `__dict__`. 

To accomplish this, a branch will need to be added to the `default_encode_value` function in `snappiershot.serializers.utils` that checks for a `__slots__` attribute and performs the encoding.

### PR DESCRIPTION
**Pull request by bbonenfant**
**Describe the Changes**
Because __slots__-optimized objects do not have an underlying dictionary, they were not supported by the `default_encode_value` serializing utility method. This PR adds support for these objects.

closes #57

### CODE REVIEW

File: snappiershot/serializers/utils.py [hunk 1]
```diff
@@ -152,5 +153,29 @@ def is_instanced_object(value: Any) -> bool:
...
+    obj_dict = copy(getattr(value, "__dict__", dict()))
+
+    # Iterate through the __slots__ (if they exist).
+    for slot in getattr(value, "__slots__", tuple()):
+        # The __dict__ object may be included in the __slots__.
+        if slot in obj_dict or slot == "__dict__":
+            continue
+        # Slots attributes are not necessarily instantiated.
+        if hasattr(value, slot):
+            obj_dict[slot] = getattr(value, slot)
```
Comment (lenabradley) [inline snappiershot/serializers/utils.py:179]: do we want a default/fallback value in the `getattr` here?

File: tests/test_serializers/test_utils.py [hunk 2]
```diff
@@ -197,6 +198,26 @@ def test_encode_numpy_recurse(value, expected) -> None:
...
+    def test_slots_class():
+        """ Test default encoding of recursive slots-optimized classes. """
+        # Arrange
+        class SlotsClass:
+            __slots__ = ("a", "b", "c", "d")
+
+            def __init__(self):
+                self.a = 1
+                self.b = 2
+                self.c = self
```
Comment (lenabradley) [inline tests/test_serializers/test_utils.py:211]: fancy fancy

File: snappiershot/serializers/utils.py [hunk 3]
```diff
@@ -152,5 +153,29 @@ def is_instanced_object(value: Any) -> bool:
     """ Check if the input value is an instanced object, i.e. an instantiated class. """
     is_type = inspect.isclass(value)
     is_function = inspect.isroutine(value)
-    is_object = hasattr(value, "__dict__")
+    is_object = hasattr(value, "__dict__") or hasattr(value, "__slots__")
     return is_object and not is_type and not is_function
+
+
+def fullvars(value: Any) -> dict:
```
Comment (lenabradley) [inline snappiershot/serializers/utils.py:160]: [nit] naming... maybe `all_vars()` or `vars_with_slots()`

File: snappiershot/serializers/utils.py [hunk 4]
```diff
@@ -152,5 +153,29 @@ def is_instanced_object(value: Any) -> bool:
...
+    """ Returns a mapping of all attributes to their associated values for a given object.
+    Supports slots-optimized classes.
+
+    Args:
+        value: An instantiated class.
+    Returns:
+        A full `vars` dictionary.
```
Comment (lenabradley) [inline snappiershot/serializers/utils.py:167]: [nit] "`vars`-style dictionary"

**Review by lenabradley — APPROVED** [review]

File: snappiershot/serializers/utils.py (hunk 1 above)
Comment (bbonenfant) [reply snappiershot/serializers/utils.py:179]: nah, no need to since entering this requires the attribute to exist

**Review by Conniemac — APPROVED** [review]
I confirmed the tests cover using slot classes

**Pull request merged**